ui/src/components/nodes/Scope.tsx

```diff
@@ -0,0 +1,247 @@
+import Grid from "@mui/material/Grid";
+import DeleteIcon from "@mui/icons-material/Delete";
+import Moveable from "react-moveable";
+
+import { useStore } from "zustand";
+
+import { RepoContext, RoleType } from "../../lib/store";
+
+interface Props {
+  data: any;
+  id: string;
+  isConnectable: boolean;
+  selected: boolean;
+  // note that xPos and yPos are the absolute position of the node
+  xPos: number;
+  yPos: number;
+}
+
+export const ScopeNode = memo<Props>(function ScopeNode({
+  data,
+  id,
+  isConnectable,
+  selected,
+}) {
+  // add resize to the node
+  const ref = useRef(null);
+  const store = useContext(RepoContext);
+  if (!store) throw new Error("Missing BearContext.Provider in the tree");
+  const reactFlowInstance = useReactFlow();
+  const setPodName = useStore(store, (state) => state.setPodName);
+  const setPodGeo = useStore(store, (state) => state.setPodGeo);
+  const [target, setTarget] = React.useState<any>();
+  const nodesMap = useStore(store, (state) => state.ydoc.getMap<Node>("pods"));
+  const [frame] = React.useState({
+    translate: [0, 0],
+  });
+  // const selected = useStore(store, (state) => state.pods[id]?.selected);
+  const role = useStore(store, (state) => state.role);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const onResize = useCallback(({ width, height, offx, offy }) => {
+    const node = nodesMap.get(id);
+    if (node) {
+      node.style = { ...node.style, width, height };
+      node.position.x += offx;
+      node.position.y += offy;
+      nodesMap.set(id, node);
+    }
+  }, []);
+
+  useEffect(() => {
+    setTarget(ref.current);
+  }, []);
+
+  useEffect(() => {
+    if (!data.name) return;
+    setPodName({ id, name: data.name || "" });
+    if (inputRef?.current) {
+      inputRef.current.value = data.name;
+    }
+  }, [data.name, id, setPodName]);
+
+  return (
+    <Box
+      ref={ref}
+      sx={{
+        width: "100%",
+        height: "100%",
+        border: "solid 1px #d6dee6",
+        borderRadius: "4px",
+      }}
+      className="custom-drag-handle"
+    >
+      <Box
+        sx={{
+          display: "flex",
+          marginLeft: "10px",
+          borderRadius: "4px",
+          position: "absolute",
+          border: "solid 1px #d6dee6",
+          right: "25px",
+          top: "-15px",
+          background: "white",
+          zIndex: 250,
+          justifyContent: "center",
+        }}
+      >
+        {role !== RoleType.GUEST && (
+          <Tooltip title="Delete" className="nodrag">
+            <IconButton
+              size="small"
+              onClick={(e: any) => {
+                // This does not work, will throw "Parent node
+                // jqgdsz2ns6k57vich0bf not found" when deleting a scope.
+                //
+                // nodesMap.delete(id);
+                //
+                // But this works:
+                reactFlowInstance.deleteElements({ nodes: [{ id }] });
```

Comment:
@li-xin-yi `nodesMap.delete` and `reactFlowInstance.deleteElements` are indeed different, te latter seems to support recursive deletion of a scope.